location-service/src/main/java/com/movie/recommender/location/service/LocationService.java

```diff
@@ -1,30 +1,86 @@
 package com.movie.recommender.location.service;
 
 import com.movie.recommender.location.exception.LocationNotFoundException;
+import com.movie.recommender.location.model.entity.GeolocationResult;
 import com.movie.recommender.location.repository.LocationRepository;
+import com.movie.recommender.location.repository.CountryRepository;
 import com.movie.recommender.location.model.dto.LocationCreateDTO;
 import com.movie.recommender.location.model.dto.LocationUpdateDTO;
+import com.movie.recommender.location.repository.CityRepository;
 import com.movie.recommender.location.model.entity.Location;
 import com.movie.recommender.location.model.dto.LocationDTO;
+import com.movie.recommender.location.model.entity.Country;
+import com.movie.recommender.location.model.entity.City;
 import org.springframework.stereotype.Service;
 import jakarta.transaction.Transactional;
 import lombok.extern.slf4j.Slf4j;
 
+
 @Slf4j
 @Service
 public class LocationService {
 
     private final LocationRepository locationRepository;
+    private final CityRepository cityRepository;
+    private final CountryRepository countryRepository;
+    private final GeolocationService geolocationService;
 
-    public LocationService(LocationRepository locationRepository) {
+    public LocationService(LocationRepository locationRepository, CityRepository cityRepository, CountryRepository countryRepository, GeolocationService geolocationService) {
         this.locationRepository = locationRepository;
+        this.cityRepository = cityRepository;
+        this.countryRepository = countryRepository;
+        this.geolocationService = geolocationService;
+    }
+
+    @Transactional
+    public City getCityByCoordinates(Double latitude, Double longitude) {
+        log.info("Fetching or creating geolocation");
+
+        GeolocationResult result;
+        result = geolocationService.fetchGeolocation(latitude, longitude);
+
+        String cityName = result.getCity();
+        String countryName = result.getCountry();
+
+        log.info("Fetching city: {} and country {} from db", cityName, countryName);
+        Country country = findOrCreateCountryByName(countryName);
+        return findOrCreateCityByName(cityName, country);
+    }
+
+    @Transactional
+    public Country findOrCreateCountryByName(String name) {
+        log.info("Finding country by name: {}", name);
+        return countryRepository.findByName(name)
+                .orElseGet(() -> {
+                    log.info("Creating new country: {}", name);
+                    Country newCountry = new Country();
```

Comment:
create a constructor with parameters to not set properties individually with setters
this can be simplified to `Country newCountry = new Country(name)`